File: app/src/main/java/vedam/subkuch/ui/chat/ChatAdapter.java
package vedam.subkuch.ui.chat;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.crashlytics.android.Crashlytics;

import java.util.ArrayList;
import java.util.List;

import vedam.subkuch.R;
import vedam.subkuch.db.chat.Chat;
import vedam.subkuch.helpers.Constants;
import vedam.subkuch.utils.AppPrefs;
import vedam.subkuch.utils.DateTimeUtils;
import vedam.subkuch.utils.UiUtil;


/**
 * Created by nadeemansari on 01/04/16.
 */
public class ChatAdapter extends RecyclerView.Adapter<ChatAdapter.ViewHolder> {

    private List<Chat> chats = new ArrayList<>();
    private Context context;

    ChatAdapter(Context context) {
        this.context = context;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(context);
        View view = layoutInflater.inflate(R.layout.fragment_chat_list_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {

        Chat chat = chats.get(position);
        int status = chat.getStatus();

        String myId = AppPrefs.getPrefsUserId(context);
        if (status == Constants.CHAT_STATUS_DELIVERED)
            holder.ivSent.setImageResource(R.drawable.ic_done_all_black_18dp);
        else if (status == Constants.CHAT_STATUS_SENT_BUT_NOT_DELIVERED) {
            holder.ivSent.setImageResource(R.drawable.ic_done_black_18dp);
        } else if (status == Constants.CHAT_STATUS_NOT_SENT) {
            holder.ivSent.setImageResource(R.drawable.baseline_schedule_black_18);
        } else {
            holder.ivSent.setImageResource(R.drawable.ic_done_all_black_18dp);
        }

        String fromId = chat.getFromProfileId();
        // else
        // img_sent.setVisibility(View.GONE);
        if (fromId.equals(myId)) {
            holder.ivSent.setVisibility(View.VISIBLE);
            holder.llAlignment.setGravity(Gravity.END);
            holder.llAlignment.setBackgroundResource(R.drawable.white_absolute);
            holder.llLayout.setGravity(Gravity.END);
            holder.llRow.setGravity(Gravity.END);
        } else {
            holder.ivSent.setVisibility(View.GONE);
            holder.llAlignment.setGravity(Gravity.START);
            holder.llAlignment.setBackgroundResource(R.drawable.white_layer);
            holder.llLayout.setGravity(Gravity.START);
            holder.llRow.setGravity(Gravity.START);
        }
        UiUtil.setTextView(holder.tvMessage, chat.getMessage());
        try {
            holder.tvMessageDate.setVisibility(View.VISIBLE);
            UiUtil.setTextView(holder.tvMessageDate,
                    DateTimeUtils.getFormattedDate(Long.parseLong(chat.getTimeStamp()), DateTimeUtils.DATE_FORMAT_4));
        } catch (NumberFormatException e) {
            Crashlytics.logException(e);
            holder.tvMessageDate.setVisibility(View.INVISIBLE);
        }
    }

    @Override
    public int getItemCount() {
        return chats.size();
    }

    static class ViewHolder extends RecyclerView.ViewHolder {

        private TextView tvMessage;
        private TextView tvMessageDate;
        LinearLayout llRow;
        LinearLayout llLayout;
        LinearLayout llAlignment;
        private ImageView ivSent;

        ViewHolder(@NonNull View itemView) {
            super(itemView);
            llAlignment = itemView.findViewById(R.id.ll_alignment);
            llLayout = itemView.findViewById(R.id.ll_container);
            llRow = itemView.findViewById(R.id.message_row_layout);
            ivSent = itemView.findViewById(R.id.chat_row_sent);
            tvMessage = itemView.findViewById(R.id.message_text);
            tvMessageDate = itemView.findViewById(R.id.message_text_date);
        }

        /*public <E> void bind(final E item, final int position, final OnListViewItemClickListener listener) {

            itemView.setOnClickListener(v -> {
                if (listener != null)
                    listener.onItemClick(item, position, itemView, null);
            });
        }*/
    }

    void setChat(List<Chat> chats) {
        this.chats = chats;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/vedam/subkuch/ui/chat/ChatAdapter.kt
package vedam.subkuch.ui.chat

import android.content.Context
import android.content.res.ColorStateList
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ListAdapter
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.core.widget.ImageViewCompat
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.google.firebase.crashlytics.FirebaseCrashlytics
import vedam.subkuch.R
import vedam.subkuch.db.chat.Chat
import vedam.subkuch.utils.AppPrefs
import vedam.subkuch.utils.DateTimeUtils
import vedam.subkuch.utils.UiUtil
import java.util.*

/**
 * Created by nadeemansari on 01/04/16.
 */
class ChatAdapter constructor(private val context: Context) :
    androidx.recyclerview.widget.ListAdapter<Chat, ChatAdapter.ViewHolder>(ChatDiffCallback()) {
    //    private var chats: List<Chat> = ArrayList()
    var myId: String = AppPrefs.getPrefsUserId(context)
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val layoutInflater = LayoutInflater.from(context)
        val view = layoutInflater.inflate(R.layout.fragment_chat_list_item, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val (_, _, fromId, _, message, timeStamp, _, status, isRead) = getItem(position)

//        if (status == Constants.CHAT_STATUS_DELIVERED)
//            holder.ivSent.setImageResource(R.drawable.ic_done_all_black_18dp);
//        else if (status == Constants.CHAT_STATUS_SENT_BUT_NOT_DELIVERED) {
//            holder.ivSent.setImageResource(R.drawable.ic_done_black_18dp);
//        } else if (status == Constants.CHAT_STATUS_NOT_SENT) {
//            holder.ivSent.setImageResource(R.drawable.baseline_schedule_black_18);
//        } else {
//            holder.ivSent.setImageResource(R.drawable.ic_done_all_black_18dp);
//        }
        if (status) {
            holder.ivSent.setImageResource(R.drawable.ic_done_black_18dp)
            ImageViewCompat.setImageTintList(
                holder.ivSent, ColorStateList.valueOf(
                    ContextCompat.getColor(context, R.color.black)
                )
            )
        } else {
            holder.ivSent.setImageResource(R.drawable.ic_done_all_black_18dp)
            if (isRead) ImageViewCompat.setImageTintList(
                holder.ivSent, ColorStateList.valueOf(
                    ContextCompat.getColor(context, R.color.blue_tick)
                )
            ) else ImageViewCompat.setImageTintList(
                holder.ivSent, ColorStateList.valueOf(
                    ContextCompat.getColor(context, R.color.black)
                )
            )
        }
        // else
        // img_sent.setVisibility(View.GONE);
        if (fromId == myId) {
            holder.ivSent.visibility = View.VISIBLE
            holder.llAlignment.gravity = Gravity.END
            holder.llAlignment.setBackgroundResource(R.drawable.white_absolute)
            holder.llLayout.gravity = Gravity.END
            holder.llRow.gravity = Gravity.END
        } else {
            holder.ivSent.visibility = View.GONE
            holder.llAlignment.gravity = Gravity.START
            holder.llAlignment.setBackgroundResource(R.drawable.white_layer)
            holder.llLayout.gravity = Gravity.START
            holder.llRow.gravity = Gravity.START
        }
        UiUtil.setTextView(holder.tvMessage, message)
        try {
            holder.tvMessageDate.visibility = View.VISIBLE
            UiUtil.setTextView(
                holder.tvMessageDate,
                DateTimeUtils.getFormattedDate(timeStamp!!.toDate().time, DateTimeUtils.DATE_FORMAT_4)
            )
        } catch (e: NumberFormatException) {
            FirebaseCrashlytics.getInstance().recordException(e)
            holder.tvMessageDate.visibility = View.INVISIBLE
        }
    }

//    override fun getItemCount(): Int {
//        return chats.size
//    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val tvMessage: TextView = itemView.findViewById(R.id.message_text)
        val tvMessageDate: TextView = itemView.findViewById(R.id.message_text_date)
        var llRow: LinearLayout = itemView.findViewById(R.id.message_row_layout)
        var llLayout: LinearLayout = itemView.findViewById(R.id.ll_container)
        var llAlignment: LinearLayout = itemView.findViewById(R.id.ll_alignment)
        val ivSent: ImageView = itemView.findViewById(R.id.chat_row_sent)

    }

    fun setChat(chats: List<Chat>) {
        submitList(chats)
//        notifyDataSetChanged()
    }

    private class ChatDiffCallback : DiffUtil.ItemCallback<Chat>() {

        override fun areItemsTheSame(oldItem: Chat, newItem: Chat): Boolean {
            return oldItem.docId == newItem.docId
        }

        override fun areContentsTheSame(oldItem: Chat, newItem: Chat): Boolean {
            return oldItem == newItem
        }
    }
}